LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/DisposableExtensions.cs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#nullable enable
+
+namespace LoRaWan.NetworkServer
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    internal static class DisposableExtensions
+    {
+        /// <remarks>
+        /// In general <see cref="IAsyncDisposable.DisposeAsync"/> implementations are not
+        /// expected to throw exceptions. If any of the disposable objects throw exceptions then
+        /// the behaviour of this method is undefined.
+        /// </remarks>
+        public static async ValueTask DisposeAllAsync(this IEnumerable<IAsyncDisposable> disposables, int concurrency)
+        {
+            ArgumentNullException.ThrowIfNull(disposables, nameof(disposables));
+            if (concurrency <= 0) throw new ArgumentOutOfRangeException(nameof(concurrency), concurrency, null);
+
+            var capacity = disposables switch
+            {
+                ICollection<IAsyncDisposable> collection => collection.Count,
+                IReadOnlyCollection<IAsyncDisposable> collection => collection.Count,
+                _ => (int?)null,
+            };
+
+            if (capacity is 0)
```

Comment:
Capacity was defined as a nullable integer before, therefore we should check
```suggestion
            if (capacity is null or 0)
```

Or am I missing something?

Reply:
If `capacity` is zero then we know `disposables` is empty. If it's `null`, it means that `disposables` is a true sequence whose count cannot be known in advance so we must iterate it. This `if` just an optimization for when we know for sure the sequence is some empty collection. Hope that clears up any smoke.

Reply:
Let's add this as comment please

Reply:
With e0f7ae9944194b7f27172b0b2ecdd925d0678611, I also added a test to cover this condition.